### PR TITLE
Fix args/kwargs modification for None.

### PR DESCRIPTION
**Pull request by cspades**
# What does this PR do ?

- 🤡 🤡 Setting `args` and `kwargs` to empty iterables when they are `None` is extremely bad practice and may corrupt any model that passes `None` to the hook. This needs to be fixed ASAP, introduced in this PR: https://github.com/NVIDIA/Megatron-LM/pull/3976

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Details

- TODO(@cspades): Ideally what our hooks should `return`...
  - If our hook modifies the arguments or output, we must return `args`, `kwargs`, `output`. We should return non-`None` values to communicate to the user that our hook modifies those arguments.
  - If our hook does not modify the arguments or output, we will return `None`. We defensively protect the original arguments, because our hook does not modify them.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 1]
```diff
@@ -728,37 +723,37 @@ def _pre_forward_param_unshard(
...
+            # Because this hook does not modify the arguments, return None.
+            # The original args and kwargs will be used.
+            return None
 
         @torch.compiler.disable
         def _register_post_backward_hook(
             post_backward_hook: callable,
             module: nn.Module,
-            args: Optional[Tuple[Any, ...]] = None,
+            args: Optional[Tuple[Any, ...]],
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Do we always have `args` here?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 2]
```diff
@@ -767,28 +762,37 @@ def _register_post_backward_hook(
...
             args = tree_unflatten(args_list, args_spec)
-            kwargs = tree_unflatten(kwargs_list, kwargs_spec)
+            if kwargs is not None:
+                kwargs_list = args_kwargs_list[len(args_list) :]
+                kwargs = tree_unflatten(kwargs_list, kwargs_spec)
 
-            # Return original input to the module forward pass.
-            return args, kwargs
+            # Return modified input to the module forward pass.
+            if args is not None and kwargs is not None:
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: `return None` should be OK?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 3]
```diff
@@ -728,37 +723,37 @@ def _pre_forward_param_unshard(
...
                 prefetch_order=PrefetchOrder.FORWARD_PASS_ORDER,
             )
-            return args, kwargs
+
+            # Because this hook does not modify the arguments, return None.
+            # The original args and kwargs will be used.
+            return None
 
         @torch.compiler.disable
         def _register_post_backward_hook(
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:725]: Do not modify the `_register_post_backward_hook` function to avoid unnecessary complex introduce.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 4]
```diff
@@ -966,9 +976,7 @@ def _register_pre_forward_param_unshard_hook(module):
             """
             if self.data_parallel_sharding_strategy != "no_shard":
                 self.forward_pre_hooks[f"{module._get_name()} parameter unshard"] = (
-                    module.register_forward_pre_hook(
-                        _pre_forward_param_unshard, prepend=True, with_kwargs=True
-                    )
+                    module.register_forward_pre_hook(_pre_forward_param_unshard, prepend=True)
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Why remove `with_kwargs`?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 1 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Yes, args is a required argument for pre- and post-forward hooks.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 2 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Yeah, all the way back to Torch 1.2.0: https://github.com/pytorch/pytorch/blob/v1.2.0/torch/nn/modules/module.py#L537

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 3 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:725]: I don't think we should call `tree_flatten(kwargs=None)` though. It's not that complicated, if we have kwargs then we add it to the end of the list, otherwise we don't consider it.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 4 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: It's not needed for this hook.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 5]
```diff
@@ -918,7 +932,15 @@ def _post_forward(module: nn.Module, input: Any, output: Any):
...
-            return output
+            # Because this hook does not modify the output, return None.
+            # The original args and kwargs will be used. This defends against
+            # situations where Megatron-FSDP hooks are called with "dummy"
+            # inputs (such as `None`), and ensures that this hook does not
+            # perform any modifications on the output.
+            # NOTE(@cspades): Such as in TransformerEngine's fused layers,
+            # where they sequentially call all sub-module hooks like this:
+            # ret = hook(submodule, None, None) for hook in hooks
+            return None
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: I still don't understand why this needs to be modified according to your statements. Please provide specific examples of TE errors.

Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: It is defensive, there are no TE errors. It means that if a user were to adopt the TE strategy of checking if a hook modifies arguments, then this would communicate to the user that "this hook does not modify the arguments".

```
                # TODO(@cspades): Loophole where if the user purposefully
                # passes None to args or kwargs, then this hook returns None.
                # This implies that the user cannot check if this hook modifies
                # args or kwargs by passing in None as a canary. Should return
                # the original args or kwargs unmodified.
```

It is what PyTorch recommends for hook design:
```
hook(module, args, kwargs) -> None or a tuple of modified input and kwargs
if hook doesn't modify arguments^         ^hook modifies arguments
```

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 3 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:725]: Fully reverted.

**Comment by cspades** [comment]
Closing this reversion in favor of just fixing-forward here: https://github.com/NVIDIA/Megatron-LM/pull/3991
